### PR TITLE
[Optim] Support tensor lr for all optimizers and check it is 1-element

### LINKED ISSUE #130980: Adam optimizer initializer useless if condition, torch v2.3.0
### 🐛 Describe the bug

There is a logical problem in Adam optimizer implementation.
It appeared when I tried to use lr as a tensor. 
1. If `lr` is a float, the second `if-condition` will not run,
2. If `lr` is a tensor, the code raises an error from the first `if-condition` and never checks the second one.

In summary, no matter what, the second condition does nothing.
One possible fix is to add `if instance(lr, float) and not 0.0 <= lr` instead.

```
class Adam(Optimizer):
    def __init__(self,
                 params: ParamsT,
                 lr: Union[float, Tensor] = 1e-3,
                 betas: Tuple[float, float] = (0.9, 0.999),
                 eps: float = 1e-8,
                 weight_decay: float = 0,
                 amsgrad: bool = False,
                 *,
                 foreach: Optional[bool] = None,
                 maximize: bool = False,
                 capturable: bool = False,
                 differentiable: bool = False,
                 fused: Optional[bool] = None):
        if not 0.0 <= lr:
            raise ValueError(f"Invalid learning rate: {lr}")
        if isinstance(lr, Tensor) and foreach and not capturable:
            raise ValueError("lr as a Tensor is not supported for capturable=False and foreach=True")
```
This is the output of the code:

Screenshot 2024-07-17 at 4 15 10 PM

### Versions

The command above did not work on my machine. However, this is the info:
`Adam optimizer initializer, torch v2.3.0, python 3.11 on macOSX (Sonoma 14.5).`

cc @vincentqb @jbschlosser @albanD @janeyx99 @crcrpar

### PR DESCRIPTION
**Pull request by qqaatw**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #131065

Fixes: #130980

### CODE REVIEW

File: torch/optim/sgd.py [hunk 1]
```diff
@@ -23,7 +23,7 @@ class SGD(Optimizer):  # noqa: D101
     def __init__(
         self,
         params,
-        lr: float = 1e-3,
+        lr: Union[float, Tensor] = 1e-3,
```
Comment (qqaatw) [inline torch/optim/sgd.py:26]: I believe Tensor LR is supported for SGD on CPU, CUDA, and MPS

**Comment by qqaatw** [comment]
Hi @janeyx99 @albanD, do you know what the reason of this failure could be?

`inductor/test_compiled_optimizers.py::CompiledOptimizerParityTestsCPU::test_correctness_LBFGS_use_closure_True_cpu_float32`

**Comment by qqaatw** [comment]
The test only fails on  Darwin, xfailed it.

File: torch/testing/_internal/common_optimizers.py [hunk 2]
```diff
@@ -1543,6 +1558,13 @@ def _get_optim_inputs_including_global_cliquey_kwargs(
                 "test_tensor_lr",
                 active_if=sys.version_info < (3, 9) and sys.version_info > (3, 7),
             ),
+            DecorateInfo(
+                unittest.expectedFailure,  # Fails if use_closure
+                "CompiledOptimizerParityTests",
+                "test_correctness",
+                active_if=lambda kwargs: sys.platform == "darwin"
```
Comment (janeyx99) [inline torch/testing/_internal/common_optimizers.py:1565]: could you open an issue to track how/why this fails?

Comment (janeyx99) [reply torch/testing/_internal/common_optimizers.py:1565]: (and then link it here)

**Review by janeyx99 — APPROVED** [review]
thanks! one comment to track the darwin failure

File: torch/testing/_internal/common_optimizers.py (hunk 2 above)
Comment (qqaatw) [reply torch/testing/_internal/common_optimizers.py:1566]: done

**Pull request closed without merging**